mayastor/src/bdev/dev/nvmx/controller_inner.rs

```diff
@@ -0,0 +1,333 @@
+}
+
+// I/O device controller API.
+impl<'a> DeviceIoController for NvmeController<'a> {
+    /// Get current I/O timeout action.
+    fn get_timeout_action(&self) -> Result<DeviceTimeoutAction, CoreError> {
+        Ok(TimeoutConfig::from_ptr(self.timeout_config).get_timeout_action())
+    }
+
+    /// Set current I/O timeout action.
+    fn set_timeout_action(
+        &mut self,
+        action: DeviceTimeoutAction,
+    ) -> Result<(), CoreError> {
+        TimeoutConfig::from_ptr(self.timeout_config).set_timeout_action(action);
+        info!("{} timeout action set to {:?}", self.name, action);
+        Ok(())
+    }
+}
+
+// I/O timeout handling for NVMe controller.
+impl<'a> NvmeController<'a> {
+    extern "C" fn command_abort_handler(
+        ctx: *mut c_void,
+        cpl: *const spdk_nvme_cpl,
+    ) {
+        let timeout_ctx = TimeoutConfig::from_ptr(ctx as *mut TimeoutConfig);
+
+        if nvme_cpl_succeeded(cpl) {
+            info!("{} CID abort succeeded for controller.", timeout_ctx.name);
+        } else {
+            error!(
+                "{} CID abort failed, resetting the controller.",
+                timeout_ctx.name
+            );
+            timeout_ctx.reset_controller();
+        }
+    }
+
+    extern "C" fn io_timeout_handler(
+        cb_arg: *mut c_void,
+        ctrlr: *mut spdk_nvme_ctrlr,
+        qpair: *mut spdk_nvme_qpair,
+        cid: u16,
+    ) {
+        let spdk_ctrlr = SpdkNvmeController::from(ctrlr);
+        let timeout_cfg = TimeoutConfig::from_ptr(cb_arg as *mut TimeoutConfig);
+        let mut timeout_action = timeout_cfg.timeout_action.load();
+
+        error!(
+            "{}: detected timeout: qpair={:p}, cid={}, action={:?}",
+            timeout_cfg.name, qpair, cid, timeout_action
+        );
+
+        // Check Controller Fatal Status for non-admin commands only to avoid
+        // endless command resubmission in case of disconnected qpair.
+        if !qpair.is_null() && spdk_ctrlr.check_cfs() {
+            error!(
+                "{}: controller Fatal Status set, reset required",
+                timeout_cfg.name
+            );
+            timeout_action = DeviceTimeoutAction::Reset;
+        }
+
+        // Handle timeout based on the action.
+        match timeout_action {
+            DeviceTimeoutAction::Abort | DeviceTimeoutAction::Reset => {
+                let mut rc = 0;
+
+                if timeout_action == DeviceTimeoutAction::Abort {
+                    error!("{}: aborting CID {}", timeout_cfg.name, cid);
+                    rc = spdk_ctrlr.abort_queued_command(
+                        qpair,
+                        cid,
+                        Some(NvmeController::command_abort_handler),
+                        cb_arg,
+                    );
+                    if rc == 0 {
+                        info!(
+                            "{}: initiated abort for CID {}",
+                            timeout_cfg.name, cid
+                        );
```

Comment:
I think it's slightly cleaner to return early here, then the next test for `timeout_action == Reset` or `rc != 0` can be removed altogether.

Reply:
Indeed the code looks shorter than. Thanks.